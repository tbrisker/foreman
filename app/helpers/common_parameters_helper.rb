module CommonParametersHelper
  # Return true if user is authorized for controller/action OR controller/action@type, otherwise false
  # third argument may be specific object (usually for edit and destroy actions)
  def authorized_via_my_scope(controller, action, object = nil)
    authorized_for(:controller => controller, :action => action, :auth_object => object)
  end

  def parameters_title
    _("Parameters that would be associated with hosts in this %s") % (type)
  end

  def global_parameter_header
    content_tag(:thead) do
      content_tag(:tr) do
        content_tag(:th, _("Name"), :class => 'col-md-3')+
          content_tag(:th, _("Value"), :class => 'col-md-7')+
          content_tag(:th, _("Actions"), :class => 'col-md-2')
      end
    end
  end

  def parameter_value_field(name, value)
    content_tag :div, :class => "input-group" do
      text_area_tag("value_#{name}", value, :rows => 1,
                    :class => "form-control", :disabled => true, :'data-hidden-value' => Parameter.hidden_value) +
        content_tag(:span, fullscreen_button('$(this).parent().prev()'), :class => 'input-group-btn')
    end
  end

  def use_puppet_default_help link_title = nil, title = _("Use Puppet default")
    popover(link_title, _("Do not send this parameter via the ENC.<br>Puppet will use the value defined in the manifest."), :title => title)
  end
end
