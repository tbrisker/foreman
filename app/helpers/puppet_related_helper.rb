module PuppetRelatedHelper
  UI.register_host_description do
    multiple_actions_provider :puppet_actions
    overview_fields_provider :puppet_host_overview_fields
    overview_buttons_provider :puppet_host_overview_buttons
    title_actions_provider :puppet_host_title_actions
  end

  def host_puppet_environment_field(form, select_options = {}, html_options = {})
    select_options = {
      :include_blank => true,
      :disable_button => _(HostsAndHostgroupsHelper::INHERIT_TEXT),
      :disable_button_enabled => inherited_by_default?(:environment_id, @host),
      :user_set => user_set?(:environment_id)}.deep_merge(select_options)

    html_options = {
      :data => {
        :url => hostgroup_or_environment_selected_hosts_path,
        :host => {
          :id => @host.id,
        },
      }}.deep_merge(html_options)

    puppet_environment_field(
      form,
      accessible_resource(@host, :environment),
      select_options,
      html_options)
  end

  def hostgroup_puppet_environment_field(form, select_options = {}, html_options = {})
    select_options = {
      :include_blank => blank_or_inherit_f(form, :environment),
    }.deep_merge(select_options)

    html_options = {
      :data => {
        :url => environment_selected_hostgroups_path,
      }}.deep_merge(html_options)

    puppet_environment_field(
      form,
      accessible_resource(@hostgroup, :environment),
      select_options,
      html_options)
  end

  def puppet_environment_field(form, environments_choice, select_options = {}, html_options = {})
    html_options = {
      :onchange => "update_puppetclasses(this)",
      :help_inline => :indicator}.deep_merge(html_options)

    select_f(
      form,
      :environment_id,
      environments_choice,
      :id,
      :to_label,
      select_options,
      html_options)
  end

  def interesting_puppetclasses(obj)
    classes = obj.all_puppetclasses
    classes_ids = classes.reorder(nil).pluck("puppetclasses.id")
    class_vars = PuppetclassLookupKey.reorder(nil).joins(:environment_classes).where(:environment_classes => { :puppetclass_id => classes_ids }).distinct.pluck("environment_classes.puppetclass_id")

    classes.where(:id => class_vars)
  end

  def puppetclasses_tab(puppetclasses_receiver)
    content_tag(:div, :class => "tab-pane", :id => "puppet_klasses") do
      if @environment.present? ||
          @hostgroup.present? && @hostgroup.environment.present?
        render "puppetclasses/class_selection", :obj => puppetclasses_receiver
      else
        alert(:class => "alert-info", :header => _("Notice"),
              :text => _("Please select an environment first"))
      end
    end
  end

  def puppet_actions
    actions = []
    if authorized_for(:controller => :hosts, :action => :edit)
      actions << { :action => [_('Change Puppet Master'), select_multiple_puppet_proxy_hosts_path], :priority => 1050 } if SmartProxy.unscoped.authorized.with_features("Puppet").exists?
      actions << { :action => [_('Change Puppet CA'), select_multiple_puppet_ca_proxy_hosts_path], :priority => 1051 } if SmartProxy.unscoped.authorized.with_features("Puppet CA").exists?
    end
    actions << { :action => [_('Run Puppet'), multiple_puppetrun_hosts_path], :priority => 1052 } if Setting[:puppetrun] && authorized_for(:controller => :hosts, :action => :puppetrun)
    actions
  end

  def puppet_host_overview_fields(host)
    fields = []
    if host.environment.present?
      fields << {
        :field => [
          _("Puppet Environment"),
          link_to(host.environment, hosts_path(:search => "environment = #{host.environment}")),
        ],
        :priority => 650,
      }
    end
    fields
  end

  def puppet_host_title_actions(host)
    if host.try(:puppet_proxy)
      [{ :action => button_group(
        link_to_if_authorized(_("Run puppet"), hash_for_puppetrun_host_path(:id => host).merge(:auth_object => host, :permission => 'puppetrun_hosts'),
                              :disabled => !Setting[:puppetrun],
                              :class => 'btn btn-default',
                              :title => _("Trigger a puppetrun on a node; requires that puppet run is enabled"))),
         :priority => 250 }]
    else
      []
    end
  end

  def puppet_host_overview_buttons(host)
    buttons = []
    buttons << { :button => link_to(_("YAML"), externalNodes_host_path(:name => host), :title => _("Puppet external nodes YAML dump"), :class => 'btn btn-default'), :priority => 400 } if SmartProxy.with_features("Puppet").any?
    buttons
  end
end
