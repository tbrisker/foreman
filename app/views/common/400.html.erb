<div class='col-md-offset-4 col-md-4'>
  <%= alert :header => (n_("Invalid parameter: %s", "Invalid parameters: %s", exception.params.length) % exception.params.to_sentence),
    :text => _('Please verify that the parameter name is valid and the values are the correct type.'),
    :actions => link_to(_('Back'), main_app.root_path, :class => 'btn btn-default'),
    :class => 'alert-warning',
    :close => false %>
</div>
