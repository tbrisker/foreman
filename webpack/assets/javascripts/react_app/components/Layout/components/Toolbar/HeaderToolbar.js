import React from 'react';
import {
  PageHeaderTools,
  PageHeaderToolsGroup,
  PageHeaderToolsItem,
} from '@patternfly/react-core';
import TaxonomySwitcher from '../TaxonomySwitcher/TaxonomySwitcher';
import UserDropdowns from './UserDropdowns';
import NotificationContainer from '../../../notifications';
import ImpersonateIcon from '../ImpersonateIcon';
import {
  layoutPropTypes,
  layoutDefaultProps,
  dataPropType,
} from '../../LayoutHelper';
import InstanceTitleViewer from './InstanceTitleViewer';
import './HeaderToolbar.scss';

const HeaderToolbar = ({
  locations,
  orgs,
  notification_url: notificationUrl,
  user,
  stop_impersonation_url: stopImpersonationUrl,
  instance_title: instanceTitle,
  isLoading,
  changeActiveMenu,
}) => (
  <PageHeaderTools id="data-toolbar">
    <PageHeaderToolsGroup className="header-tool-item-hidden-lg">
      <TaxonomySwitcher
        locations={locations.available_locations || []}
        organizations={orgs.available_organizations || []}
        isLoading={isLoading}
      />
    </PageHeaderToolsGroup>
    <PageHeaderToolsGroup>
      <PageHeaderToolsItem>
        <InstanceTitleViewer title={instanceTitle} />
      </PageHeaderToolsItem>
      <PageHeaderToolsItem className="notifications_container">
        <NotificationContainer data={{ url: notificationUrl }} />
      </PageHeaderToolsItem>
      {user.impersonated_by && (
        <PageHeaderToolsItem className="impersonation-item">
          <ImpersonateIcon stopImpersonationUrl={stopImpersonationUrl} />
        </PageHeaderToolsItem>
      )}

      <PageHeaderToolsItem className="header-tool-item-hidden-lg user-nav-item">
        <UserDropdowns
          notificationUrl={notificationUrl}
          user={user}
          changeActiveMenu={changeActiveMenu}
        />
      </PageHeaderToolsItem>
    </PageHeaderToolsGroup>
  </PageHeaderTools>
);
HeaderToolbar.propTypes = {
  ...dataPropType,
  isLoading: layoutPropTypes.isLoading,
  changeActiveMenu: layoutPropTypes.changeActiveMenu,
};

HeaderToolbar.defaultProps = {
  isLoading: layoutDefaultProps.isLoading,
  changeActiveMenu: layoutDefaultProps.changeActiveMenu,
};
export default HeaderToolbar;
