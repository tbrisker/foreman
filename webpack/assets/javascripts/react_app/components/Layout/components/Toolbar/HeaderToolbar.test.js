import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import { hasTaxonomiesMock } from '../../Layout.fixtures';
import { noop } from '../../../../common/helpers';

import HeaderToolbar from './HeaderToolbar';

const fixtures = {
  'render HeaderToolbar': {
    ...hasTaxonomiesMock.data,
    currentLocation: hasTaxonomiesMock.currentLocation,
    currentOrganization: hasTaxonomiesMock.currentOrganization,
    isLoading: false,
    changeActiveMenu: noop,
  },
};

describe('HeaderToolbar', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(HeaderToolbar, fixtures));
});
