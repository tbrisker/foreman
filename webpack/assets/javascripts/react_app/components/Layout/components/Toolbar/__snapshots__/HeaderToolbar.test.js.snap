// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderToolbar rendering render HeaderToolbar 1`] = `
<PageHeaderTools
  id="data-toolbar"
>
  <PageHeaderToolsGroup
    className="header-tool-item-hidden-lg"
  >
    <TaxonomySwitcher
      isLoading={false}
      locations={
        Array [
          Object {
            "href": "/locations/1-yaml/select",
            "id": 1,
            "title": "yaml",
          },
          Object {
            "href": "/locations/2-london/select",
            "id": 2,
            "title": "london",
          },
          Object {
            "href": "/locations/3-norway/select",
            "id": 3,
            "title": "norway",
          },
        ]
      }
      organizations={
        Array [
          Object {
            "href": "/organizations/1-org1/select",
            "id": 1,
            "title": "org1",
          },
          Object {
            "href": "/organizations/2-org2/select",
            "id": 2,
            "title": "org2",
          },
        ]
      }
    />
  </PageHeaderToolsGroup>
  <PageHeaderToolsGroup>
    <PageHeaderToolsItem>
      <InstanceTitleViewer
        title="Production"
      />
    </PageHeaderToolsItem>
    <PageHeaderToolsItem
      className="notifications_container"
    >
      <Connect(OnClickOutside(notificationContainer))
        data={
          Object {
            "url": "/notification_recipients",
          }
        }
      />
    </PageHeaderToolsItem>
    <PageHeaderToolsItem
      className="impersonation-item"
    >
      <Connect(ImpersonateIcon)
        stopImpersonationUrl="/users/stop_impersonation"
      />
    </PageHeaderToolsItem>
    <PageHeaderToolsItem
      className="header-tool-item-hidden-lg user-nav-item"
    >
      <UserDropdowns
        changeActiveMenu={[Function]}
        className=""
        instanceTitle=""
        notificationUrl="/notification_recipients"
        user={
          Object {
            "current_user": Object {
              "firstname": "Admin",
              "id": 4,
              "lastname": "User",
              "login": "admin",
              "name": "Admin User",
            },
            "impersonated_by": true,
            "user_dropdown": Array [
              Object {
                "children": Array [
                  Object {
                    "isDivider": false,
                    "onClick": [MockFunction],
                    "title": "Aa",
                  },
                  Object {
                    "isDivider": false,
                    "onClick": [MockFunction],
                    "title": "Cc",
                  },
                ],
                "icon": "fa fa-user",
                "name": "User",
                "type": "sub_menu",
              },
            ],
          }
        }
      />
    </PageHeaderToolsItem>
  </PageHeaderToolsGroup>
</PageHeaderTools>
`;
